File: cotta-guice/src/main/java/com/mgtriffid/games/cotta/experimental/guice/beans/impl/ServerSimulationImpl.java
package com.mgtriffid.games.cotta.experimental.guice.beans.impl;

import com.mgtriffid.games.cotta.experimental.guice.beans.EffectBus;
import com.mgtriffid.games.cotta.experimental.guice.beans.PlayerIdGenerator;
import com.mgtriffid.games.cotta.experimental.guice.beans.ServerSimulation;
import com.mgtriffid.games.cotta.experimental.guice.beans.SimulationTick;
import com.mgtriffid.games.cotta.experimental.guice.beans.state.JCottaState;
import jakarta.inject.Inject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ServerSimulationImpl implements ServerSimulation {
    private final JCottaState state;
    private final List<Class> metaEntitiesInputComponents;
    private final PlayerIdGenerator playerIdGenerator;
    private final SimulationTick tick;
    private final EffectBus effectBus;
    private final Map<Integer, Map<String, String>> enterGameIntents = new HashMap<>();

    public ServerSimulationImpl(
            JCottaState state,
            List<Class> metaEntitiesInputComponents,
            PlayerIdGenerator playerIdGenerator,
            SimulationTick tick,
            EffectBus effectBus
    ) {
        this.state = state;
        this.metaEntitiesInputComponents = new ArrayList(metaEntitiesInputComponents);
        this.playerIdGenerator = playerIdGenerator;
        this.tick = tick;
        this.effectBus = effectBus;
    }

    @Override
    public void registerSystem(Class system) {

    }

    @Override
    public int enterGame(Map<String, String> params) {
        int playerId = playerIdGenerator.id();
        enterGameIntents.put(playerId, params);
        return playerId;
    }

    @Override
    public void tick() {
        effectBus.clear();
        state.advance(tick.tick);
        tick.tick++;
        putInputIntoEntities();
        invokeSystems();
        processEnterGameIntents();
    }

    private void putInputIntoEntities() {
        state.getEntities(); // and we put input in
    }
    private void invokeSystems() {

    }

    private void processEnterGameIntents() {

    }
}

File: cotta-server/src/main/kotlin/com/mgtriffid/games/cotta/server/impl/ServerSimulationImpl.kt
package com.mgtriffid.games.cotta.server.impl

import com.google.inject.Inject
import com.mgtriffid.games.cotta.core.effects.EffectBus
import com.mgtriffid.games.cotta.core.entities.CottaState
import com.mgtriffid.games.cotta.core.entities.PlayerId
import com.mgtriffid.games.cotta.core.entities.TickProvider
import com.mgtriffid.games.cotta.core.input.InputProcessing
import com.mgtriffid.games.cotta.core.simulation.PlayersSawTicks
import com.mgtriffid.games.cotta.core.simulation.SimulationInput
import com.mgtriffid.games.cotta.core.simulation.invokers.InvokersFactory
import com.mgtriffid.games.cotta.core.simulation.invokers.SystemInvoker
import com.mgtriffid.games.cotta.core.systems.CottaSystem
import com.mgtriffid.games.cotta.network.purgatory.EnterGameIntent
import com.mgtriffid.games.cotta.server.ServerSimulation
import jakarta.inject.Named
import mu.KotlinLogging
import java.util.concurrent.atomic.AtomicInteger
import kotlin.reflect.KClass

private val logger = KotlinLogging.logger {}

class ServerSimulationImpl @Inject constructor(
    @Named("simulation") private val state: CottaState,
    private val invokersFactory: InvokersFactory,
    private val effectBus: EffectBus,
    private val playersSawTicks: PlayersSawTicks,
    private val tickProvider: TickProvider,
    private val inputProcessing: InputProcessing
) : ServerSimulation {
    private val systemInvokers = ArrayList<Pair<SystemInvoker<*>, CottaSystem>>()

    private val enterGameIntents = ArrayList<Pair<EnterGameIntent, PlayerId>>()
    private val playerIdGenerator = PlayerIdGenerator()

    override fun <T : CottaSystem> registerSystem(systemClass: KClass<T>) {
        logger.info { "Registering system '${systemClass.simpleName}'" }
        systemInvokers.add(invokersFactory.createInvoker(systemClass))
    }

    override fun tick(input: SimulationInput) {
        effectBus.clear()
        state.advance(tickProvider.tick)
        logger.debug { "Advancing tick from ${tickProvider.tick} to ${tickProvider.tick + 1}" }
        tickProvider.tick++
        processInput(input)
        fillPlayersSawTicks(input)
        simulate()
        processEnterGameIntents()
    }

    private fun processInput(input: SimulationInput) {
        logger.debug { "Processing input on server on tick ${tickProvider.tick}: ${input.inputForPlayers()}" }
        inputProcessing.process(
            input,
            state.entities(tickProvider.tick),
            effectBus
        )
    }

    private fun simulate() {
        for ((invoker, system) in systemInvokers) {
            (invoker as SystemInvoker<CottaSystem>).invoke(system) // TODO cast issue
        }
    }

    private fun fillPlayersSawTicks(input: SimulationInput) {
        playersSawTicks.set(input.playersSawTicks())
    }

    override fun enterGame(intent: EnterGameIntent): PlayerId {
        val playerId = playerIdGenerator.nextId()
        enterGameIntents.add(Pair(intent, playerId))
        return playerId
    }

    private fun processEnterGameIntents() {
        logger.debug { "Processing enterGameIntents" }
        enterGameIntents.forEach {
            logger.debug { "Processing intent to ETG for ${it.second}" }
        }
        enterGameIntents.clear()
    }

    private class PlayerIdGenerator {
        val counter = AtomicInteger(0)

        fun nextId(): PlayerId = PlayerId(counter.incrementAndGet())
    }
}
